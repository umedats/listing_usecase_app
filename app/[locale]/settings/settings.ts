/**
 * Object which will store the necessary settings required to make SP API calls.
 * Each field in this object corresponds to a field in the settings page.
 *
 * region: AWS Region for creating infrastructure
 * accountId: AWS AccountId for creating infrastructure
 * clientId: Client Id part of the selling partner application LWA credentials
 * clientSecret: Client Secret part of the selling partner application LWA credentials
 * refreshToken: Refresh Token used to get the access token
 * sellingPartnerId: Selling Partner Id
 * sellingPartnerIdType: A selling partner identifier type
 * marketplaceId: The MarketplaceId to be used during the Listing Process
 */
export type Settings = {
  region: string;
  accountId: string;
  clientId: string;
  clientSecret: string;
  refreshToken: string;
  sellingPartnerId: string;
  sellingPartnerIdType: string;
  marketplaceId: string;
};

/**
 * returns true if the settings are incomplete.
 * @param settings
 */
export function areSettingsIncomplete(settings: Settings) {
  return (
    !settings.region ||
    !settings.accountId ||
    !settings.clientId ||
    !settings.clientSecret ||
    !settings.refreshToken ||
    !settings.sellingPartnerId ||
    !settings.sellingPartnerIdType ||
    !settings.marketplaceId
  );
}

export const SELLING_PARTNER_TYPE_KEY_VENDOR_CODE = "Vendor Code";
export const SELLING_PARTNER_TYPE_KEY_MERCHANT_ACCOUNT_ID =
  "Merchant Account Id";
