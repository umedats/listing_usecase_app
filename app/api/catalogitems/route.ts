import { NextResponse } from "next/server";
import { headers } from "next/dist/client/components/headers";
import getSPAPIEndpoint, {
  buildCatalogItemsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import {
  IDENTIFIER_TYPES_HEADER,
  IDENTIFIERS_HEADER,
  KEYWORDS_HEADER,
  SEARCH_CATALOG_ITEMS_API_DOC_LINK,
  SEARCH_CATALOG_ITEMS_API_NAME,
  LOCALE_HEADER,
} from "@/app/constants/global";
import nextResponse from "@/app/utils/next-response-factory";
import { Item, ItemSearchResults } from "@/app/sdk/catalogItems_2022-04-01";
import { ProductSearchResult } from "@/app/[locale]/create-offer/product-result-type";
import { Settings } from "@/app/[locale]/settings/settings";
import { SPAPIRequestResponse } from "@/app/model/types";
import { serializeToJsonString } from "@/app/utils/serialization";
import {
  handleSPAPIInvocationError,
  retrieveSettingsAndInvokeAPILogic,
} from "@/app/utils/api";

/**
 * This export forces the NextJs not to pre-render this api route.
 */
export const dynamic = "force-dynamic";

/**
 * Retrieves the list of Catalog Items based on identifiers/keywords by calling the SearchCatalogItems SP-API.
 * @constructor
 */
export async function GET() {
  return retrieveSettingsAndInvokeAPILogic(getCatalogItemsHandler);
}

async function getCatalogItemsHandler(settings: Settings) {
  const { identifiersType, identifiers, keywords, locale } =
    retrieveRequestParamsFromHeader();
  const optionalParams = constructOptionalParametersForApi(
    settings,
    identifiersType,
    identifiers,
    keywords,
    locale,
  );
  const nextResponseOrSPAPIRequestResponse = await fetchCatalogItems(
    settings,
    optionalParams,
  );

  if (nextResponseOrSPAPIRequestResponse instanceof NextResponse) {
    return nextResponseOrSPAPIRequestResponse;
  }

  return fetchItemSearchResults(nextResponseOrSPAPIRequestResponse);
}

function retrieveRequestParamsFromHeader() {
  const requestHeaders = headers();
  const identifiersType = requestHeaders.get(IDENTIFIER_TYPES_HEADER);
  const identifierData = requestHeaders.get(IDENTIFIERS_HEADER);
  const identifiers = identifierData ? identifierData.split(",") : null;
  const keywordsData = requestHeaders.get(KEYWORDS_HEADER);
  const keywords = keywordsData ? keywordsData.split(",") : null;
  const locale = requestHeaders.get(LOCALE_HEADER);

  return { identifiersType, identifiers, keywords, locale };
}

function constructOptionalParametersForApi(
  settings: Settings,
  identifiersType: string | null,
  identifiers: string[] | null,
  keywords: string[] | null,
  locale: string | null,
) {
  const includedData = ["productTypes", "summaries"];

  if (identifiers) {
    const optionalParametersForIdentifiers = {
      identifiers: identifiers,
      identifiersType: identifiersType,
      includedData: includedData,
      locale: locale,
    };

    // Seller ID is mandatory when identifier Type is SKU
    if (identifiersType === "SKU") {
      return {
        sellerId: settings.sellingPartnerId,
        ...optionalParametersForIdentifiers,
      };
    }

    // For non-SKU Identifier Type use case.
    return optionalParametersForIdentifiers;
  }

  if (keywords) {
    return {
      keywords: keywords,
      includedData: includedData,
      locale: locale,
    };
  }
}

async function fetchCatalogItems(settings: Settings, opts: any) {
  // Build Catalog Items API Client.
  const catalogApi = await buildCatalogItemsAPIClient(
    getSPAPIEndpoint(settings.region),
  );

  const reqResponse = {
    apiName: SEARCH_CATALOG_ITEMS_API_NAME,
    apiDocumentationLink: SEARCH_CATALOG_ITEMS_API_DOC_LINK,
    request: {
      marketplaceIds: [settings.marketplaceId],
      ...opts,
    },
    response: {},
  };
  try {
    reqResponse.response = await catalogApi.searchCatalogItems(
      reqResponse.request.marketplaceIds,
      opts,
    );
    return reqResponse;
  } catch (error: any) {
    return handleSPAPIInvocationError(reqResponse, error);
  }
}

function deriveTitle(item: Item) {
  return item.summaries?.[0]?.itemName;
}

function derivePTD(item: Item) {
  return item.productTypes?.[0]?.productType;
}

function fetchItemSearchResults(reqResponse: SPAPIRequestResponse) {
  const searchResults: ItemSearchResults =
    ItemSearchResults.constructFromObject(reqResponse.response, undefined);

  if (searchResults.numberOfResults && searchResults.numberOfResults > 0) {
    // @ts-ignore
    const items: Item[] = searchResults.items;
    const results: ProductSearchResult[] = items.map((item, index) => {
      const searchResult: ProductSearchResult = {
        title: deriveTitle(item),
        asin: item.asin,
        productType: derivePTD(item),
      };
      return searchResult;
    });
    return nextResponse(
      200,
      "OK",
      serializeToJsonString({
        data: results,
        debugContext: [reqResponse],
      }),
    );
  }

  return nextResponse(
    200,
    "OK",
    serializeToJsonString({
      data: [],
      debugContext: [reqResponse],
    }),
  );
}
