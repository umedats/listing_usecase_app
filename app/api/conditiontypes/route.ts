import nextResponse from "@/app/utils/next-response-factory";
import { Settings } from "@/app/[locale]/settings/settings";
import { NextResponse } from "next/server";
import { headers } from "next/dist/client/components/headers";
import {
  ASIN_HEADER,
  GET_LISTING_RESTRICTIONS_API_DOC_LINK,
  GET_LISTING_RESTRICTIONS_API_NAME,
  LOCALE_HEADER,
} from "@/app/constants/global";
import getSPAPIEndpoint, {
  buildListingsRestrictionsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import {
  Restriction,
  RestrictionList,
} from "@/app/sdk/listingsRestrictions_2021-08-01";
import { serializeToJsonString } from "@/app/utils/serialization";
import {
  handleSPAPIInvocationError,
  retrieveSettingsAndInvokeAPILogic,
} from "@/app/utils/api";

/**
 * This export forces the NextJs not to pre-render this api route.
 */
export const dynamic = "force-dynamic";

/**
 * Retrieves the conditions under which a MCID can list a given ASIN by calling SPAPI.
 */
export async function GET() {
  return retrieveSettingsAndInvokeAPILogic(getConditionTypesHandler);
}

async function getConditionTypesHandler(settings: Settings) {
  // Validate the various headers expected by the API.
  const validateHeadersResponse = validateHeaderParams();
  if (validateHeadersResponse instanceof NextResponse) {
    return validateHeadersResponse;
  }

  const { asin, locale } = validateHeadersResponse;

  // Retrieve restricted Conditions from SP API
  const nextResponseOrSPAPIResponse = await fetchRestrictions(
    settings,
    asin,
    locale,
  );
  if (nextResponseOrSPAPIResponse instanceof NextResponse) {
    return nextResponseOrSPAPIResponse;
  }
  const restrictedConditionTypes: string[] =
    RestrictionList.constructFromObject(
      nextResponseOrSPAPIResponse.response,
      undefined,
    ).restrictions.map((r: Restriction) => r.conditionType);

  const allowedTypes = Object.values(Restriction.ConditionTypeEnum).filter(
    (c) => !restrictedConditionTypes.includes(c),
  );
  return nextResponse(
    200,
    "OK",
    serializeToJsonString({
      data: allowedTypes,
      debugContext: [nextResponseOrSPAPIResponse],
    }),
  );
}

// Validates the header parameters.
function validateHeaderParams() {
  const requestHeaders = headers();
  const asin = requestHeaders.get(ASIN_HEADER);
  if (!asin) {
    const errorMessage = "The request is missing the asin in the headers";
    console.log(errorMessage);
    return nextResponse(400, errorMessage);
  }

  const locale = requestHeaders.get(LOCALE_HEADER);

  return { asin, locale };
}

/**
 * fetches the condition types from the SP API.
 * @param settings user entered settings
 * @param asin asin
 * @param locale user preferred locale.
 */
async function fetchRestrictions(
  settings: Settings,
  asin: string,
  locale: string | null,
) {
  // Build the Restrictions API Client.
  const restrictionsApi = await buildListingsRestrictionsAPIClient(
    getSPAPIEndpoint(settings.region),
  );

  const reqResponse = {
    apiName: GET_LISTING_RESTRICTIONS_API_NAME,
    apiDocumentationLink: GET_LISTING_RESTRICTIONS_API_DOC_LINK,
    request: {
      asin: asin,
      sellerId: settings.sellingPartnerId,
      marketplaceIds: [settings.marketplaceId],
      conditionType: null,
      reasonLocale: locale ? locale : "",
    },
    response: {},
  };

  try {
    reqResponse.response = await restrictionsApi.getListingsRestrictions(
      reqResponse.request.asin,
      reqResponse.request.sellerId,
      reqResponse.request.marketplaceIds,
      {
        conditionType: reqResponse.request.conditionType,
        reasonLocale: reqResponse.request.reasonLocale,
      },
    );
    return reqResponse;
  } catch (error: any) {
    return handleSPAPIInvocationError(reqResponse, error);
  }
}
