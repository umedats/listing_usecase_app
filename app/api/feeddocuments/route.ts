import nextResponse from "@/app/utils/next-response-factory";
import { serializeToJsonString } from "@/app/utils/serialization";
import { NextResponse } from "next/server";
import { Settings } from "@/app/[locale]/settings/settings";
import { headers } from "next/dist/client/components/headers";
import {
  CONTENT_TYPE_HEADER,
  FEED_DOCUMENT_ID_HEADER,
  GET_FEED_DOCUMENT_API_NAME,
  GET_FEED_DOCUMENT_DOC_LINK,
  NO_CACHING_DIRECTIVE,
} from "@/app/constants/global";
import getSPAPIEndpoint, {
  buildFeedsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import { FeedDocument } from "@/app/sdk/feeds_2021-06-30";
import {
  handleSPAPIInvocationError,
  retrieveSettingsAndInvokeAPILogic,
} from "@/app/utils/api";
import { unzipSync } from "zlib";
import { SPAPIRequestResponse } from "@/app/model/types";

/**
 * This export forces the NextJs not to pre-render this api route.
 */
export const dynamic = "force-dynamic";

/**
 * Retrieves a feed document content and returns the content in the API response.
 * @constructor
 */
export async function GET() {
  return retrieveSettingsAndInvokeAPILogic(getFeedDocumentContentHandler);
}

async function getFeedDocumentContentHandler(settings: Settings) {
  const requiredParametersResponse =
    await validateAndRetrieveRequiredParametersForGetRequest();

  if (requiredParametersResponse instanceof NextResponse) {
    return requiredParametersResponse;
  }

  const { feedDocumentId } = requiredParametersResponse;

  const nextResponseOrSPAPIRequestResponse = await getFeedDocumentUrl(
    settings,
    feedDocumentId,
  );

  if (nextResponseOrSPAPIRequestResponse instanceof NextResponse) {
    return nextResponseOrSPAPIRequestResponse;
  }

  const getFeedDocumentContentResponse = await getFeedDocumentContent(
    FeedDocument.constructFromObject(
      nextResponseOrSPAPIRequestResponse.response,
      undefined,
    ),
    nextResponseOrSPAPIRequestResponse,
  );

  if (getFeedDocumentContentResponse instanceof NextResponse) {
    return getFeedDocumentContentResponse;
  }

  return nextResponse(
    200,
    "OK",
    serializeToJsonString({
      data: {
        content: getFeedDocumentContentResponse.feedDocumentContent,
        contentType: getFeedDocumentContentResponse.feedDocumentContentType,
      },
      debugContext: [nextResponseOrSPAPIRequestResponse],
    }),
  );
}

async function validateAndRetrieveRequiredParametersForGetRequest() {
  const validateHeadersResponse =
    validateAndRetrieveRequestParamsFromHeaderForGetRequest();
  if (validateHeadersResponse instanceof NextResponse) {
    return validateHeadersResponse;
  }

  const { feedDocumentId } = validateHeadersResponse;
  return { feedDocumentId };
}

function validateAndRetrieveRequestParamsFromHeaderForGetRequest() {
  const requestHeaders = headers();
  const feedDocumentId = requestHeaders.get(FEED_DOCUMENT_ID_HEADER);
  if (!feedDocumentId) {
    const errorMessage =
      "The request is missing the feedDocumentId in the headers";
    return nextResponse(400, errorMessage);
  }

  return { feedDocumentId };
}

async function getFeedDocumentUrl(settings: Settings, feedDocumentId: string) {
  const feedsApi = await buildFeedsAPIClient(getSPAPIEndpoint(settings.region));
  const reqResponse = {
    apiName: GET_FEED_DOCUMENT_API_NAME,
    apiDocumentationLink: GET_FEED_DOCUMENT_DOC_LINK,
    request: {
      feedDocumentId: [feedDocumentId],
    },
    response: {},
  };

  try {
    reqResponse.response = await feedsApi.getFeedDocument(feedDocumentId);
    return reqResponse;
  } catch (error: any) {
    return handleSPAPIInvocationError(reqResponse, error);
  }
}

async function getFeedDocumentContent(
  feedDocument: FeedDocument,
  requestResponse: SPAPIRequestResponse,
) {
  try {
    const fetchDocumentResponse = await fetch(feedDocument.url, {
      cache: NO_CACHING_DIRECTIVE,
    });
    const feedDocumentContentType =
      fetchDocumentResponse.headers.get(CONTENT_TYPE_HEADER);
    let feedDocumentContent = "";
    if (feedDocument.compressionAlgorithm) {
      feedDocumentContent = unzipSync(
        await fetchDocumentResponse.arrayBuffer(),
      ).toString();
    } else {
      feedDocumentContent = await fetchDocumentResponse.text();
    }
    return { feedDocumentContent, feedDocumentContentType };
  } catch (error) {
    // Generally these errors are recoverable.
    const errorMessage = "Error while fetching the feed document content.";
    console.log(errorMessage + " " + error);
    return nextResponse(
      500,
      errorMessage,
      serializeToJsonString({
        debugContext: [requestResponse],
      }),
    );
  }
}
