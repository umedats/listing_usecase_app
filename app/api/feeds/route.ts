import nextResponse from "@/app/utils/next-response-factory";
import { NextRequest, NextResponse } from "next/server";
import { serializeToJsonString } from "@/app/utils/serialization";
import {
  CREATE_FEED_API_DOC_LINK,
  CREATE_FEED_API_NAME,
  CREATE_FEED_DOCUMENT_API_DOC_LINK,
  CREATE_FEED_DOCUMENT_API_NAME,
  GET_FEEDS_API_NAME,
  GET_FEEDS_DOC_LINK,
  JSON_FILE_TYPE,
  JSON_LISTINGS_FEED_TYPE,
  NO_CACHING_DIRECTIVE,
} from "@/app/constants/global";
import { Settings } from "@/app/[locale]/settings/settings";
import getSPAPIEndpoint, {
  buildFeedsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import {
  CreateFeedDocumentResponse,
  CreateFeedDocumentSpecification,
  CreateFeedSpecification,
  GetFeedsResponse,
} from "@/app/sdk/feeds_2021-06-30";
import {
  handleChainedSPAPIInvocationError,
  handleSPAPIInvocationError,
  retrieveSettingsAndInvokeAPILogic,
  retrieveSettingsAndInvokeAPILogicWithRequest,
} from "@/app/utils/api";

/**
 * This export forces the NextJs not to pre-render this api route.
 */
export const dynamic = "force-dynamic";

/**
 * Creates a feed and return feed details in the API Response.
 * @param request the next request.
 * @constructor
 */
export async function POST(request: NextRequest) {
  return retrieveSettingsAndInvokeAPILogicWithRequest(
    createFeedHandler,
    request,
  );
}

async function createFeedHandler(settings: Settings, request: NextRequest) {
  const formData = await request.blob();

  const debugContext: any[] = [];

  const selectedFileContent = await formData.text();
  // Create and upload Document
  const nextOrSPAPIRequestResponseForDocumentCreate =
    await fetchCreateFeedDocument(settings, debugContext, selectedFileContent);

  if (nextOrSPAPIRequestResponseForDocumentCreate instanceof NextResponse) {
    return nextOrSPAPIRequestResponseForDocumentCreate;
  }

  const feedDocumentResponse =
    nextOrSPAPIRequestResponseForDocumentCreate.response as CreateFeedDocumentResponse;

  // Create Feed
  const nextOrSPAPIResponseForCreateFeed = await fetchCreateFeed(
    settings,
    feedDocumentResponse,
    debugContext,
  );
  if (nextOrSPAPIResponseForCreateFeed instanceof NextResponse) {
    return nextOrSPAPIResponseForCreateFeed;
  }

  return nextResponse(
    200,
    "OK",
    serializeToJsonString({
      data: nextOrSPAPIResponseForCreateFeed.response,
      debugContext: debugContext,
    }),
  );
}

/**
 * API which returns the past submitted feeds to the UI. Only the first page of
 * results are sent back.
 */
export async function GET() {
  return retrieveSettingsAndInvokeAPILogic(getPastFeedsHandler);
}

async function getPastFeedsHandler(settings: Settings) {
  const nextResponseOrSPAPIRequestResponse = await fetchFeeds(settings);

  if (nextResponseOrSPAPIRequestResponse instanceof NextResponse) {
    return nextResponseOrSPAPIRequestResponse;
  }

  const getFeedsResponse = GetFeedsResponse.constructFromObject(
    nextResponseOrSPAPIRequestResponse.response,
    undefined,
  );
  const feeds: any[] = getFeedsResponse?.feeds?.length
    ? getFeedsResponse.feeds
    : [];

  return nextResponse(
    200,
    "OK",
    serializeToJsonString({
      data: feeds,
      debugContext: [nextResponseOrSPAPIRequestResponse],
    }),
  );
}

async function fetchCreateFeedDocument(
  settings: Settings,
  debugContext: any[],
  feedContent: string,
) {
  const feedsApi = await buildFeedsAPIClient(getSPAPIEndpoint(settings.region));
  const jsonFeedDocSpec: CreateFeedDocumentSpecification =
    new CreateFeedDocumentSpecification(JSON_FILE_TYPE);
  const reqResponse = {
    apiName: CREATE_FEED_DOCUMENT_API_NAME,
    apiDocumentationLink: CREATE_FEED_DOCUMENT_API_DOC_LINK,
    request: jsonFeedDocSpec,
    response: {},
  };

  try {
    // Create Document
    const response: CreateFeedDocumentResponse =
      await feedsApi.createFeedDocument(jsonFeedDocSpec);

    // Upload Document
    await fetch(response.url, {
      cache: NO_CACHING_DIRECTIVE,
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: feedContent,
    });
    reqResponse.response = response;
    debugContext.push(reqResponse);
    return reqResponse;
  } catch (error: any) {
    return handleChainedSPAPIInvocationError(error, debugContext, reqResponse);
  }
}

async function fetchCreateFeed(
  settings: Settings,
  createFeedDocumentResponse: CreateFeedDocumentResponse,
  debugContext: any[],
) {
  const feedsApi = await buildFeedsAPIClient(getSPAPIEndpoint(settings.region));
  const feedRequest: CreateFeedSpecification = new CreateFeedSpecification(
    JSON_LISTINGS_FEED_TYPE,
    [settings.marketplaceId],
    createFeedDocumentResponse.feedDocumentId,
  );
  const reqResponse = {
    apiName: CREATE_FEED_API_NAME,
    apiDocumentationLink: CREATE_FEED_API_DOC_LINK,
    request: feedRequest,
    response: {},
  };

  try {
    reqResponse.response = await feedsApi.createFeed(feedRequest);
    debugContext.push(reqResponse);
    return reqResponse;
  } catch (error: any) {
    return handleChainedSPAPIInvocationError(error, debugContext, reqResponse);
  }
}

async function fetchFeeds(settings: Settings) {
  const feedsApi = await buildFeedsAPIClient(getSPAPIEndpoint(settings.region));
  const reqResponse = {
    apiName: GET_FEEDS_API_NAME,
    apiDocumentationLink: GET_FEEDS_DOC_LINK,
    request: {
      feedTypes: [JSON_LISTINGS_FEED_TYPE],
      marketplaceIds: [settings.marketplaceId],
      pageSize: 20,
    },
    response: {},
  };

  try {
    reqResponse.response = await feedsApi.getFeeds(reqResponse.request);
    return reqResponse;
  } catch (error: any) {
    return handleSPAPIInvocationError(reqResponse, error);
  }
}
