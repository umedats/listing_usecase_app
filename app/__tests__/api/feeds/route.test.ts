import getSPAPIEndpoint, {
  buildFeedsAPIClient,
} from "@/app/sdk/factory/ListingsSPAPIClientsFactory";
import { getSettings } from "@/app/api/settings/wrapper";
import fetch from "jest-fetch-mock";
import { MOCK_SETTINGS } from "@/app/test-utils/mock-settings";
import { verifyStatusAndBody } from "@/app/test-utils/next-response-helper";
import { FeedsApi } from "@/app/sdk/feeds_2021-06-30";
import { GET, POST } from "@/app/api/feeds/route";
import {
  MOCK_FEED_1,
  MOCK_FEED_2,
  MOCK_CREATE_FEED_DOCUMENT_SUCCESS,
  MOCK_CREATE_FEED_SUCCESS,
} from "@/app/test-utils/mock-feed";
import { JSON_FILE_TYPE } from "@/app/constants/global";
import { NextRequest } from "next/server";
import {
  mockRejectFetchResponse,
  mockResolveFetchResponse,
} from "@/app/test-utils/mock-fetch";

const MOCK_SPI_API_URL: string = "https://sellingpartnerapi-na.amazon.com";

const MOCK_GET_FEEDS_RESPONSE = {
  feeds: [MOCK_FEED_1, MOCK_FEED_2],
};

const feedsApi: FeedsApi = new FeedsApi(undefined);
const mockedGetFeeds = jest.fn();
const mockedCreateFeedDocument = jest.fn();
const mockedCreateFeed = jest.fn();

feedsApi.getFeeds = mockedGetFeeds;
feedsApi.createFeedDocument = mockedCreateFeedDocument;
feedsApi.createFeed = mockedCreateFeed;

function buildPostNextRequest() {
  const feedContentBlob = new Blob(["{   }"], {
    type: JSON_FILE_TYPE,
  });
  return new NextRequest("http://localhost:3000/api/feeds", {
    method: "POST",
    body: feedContentBlob,
  });
}

function mockCreateFeedDocument(error?: object) {
  mockedGetSPAPIEndpoint.mockReturnValue(MOCK_SPI_API_URL);
  if (error) {
    mockedCreateFeedDocument.mockRejectedValue(error);
  } else {
    mockedCreateFeedDocument.mockResolvedValue(
      MOCK_CREATE_FEED_DOCUMENT_SUCCESS,
    );
  }
}

function mockCreateFeed(error?: object) {
  mockedGetSPAPIEndpoint.mockReturnValue(MOCK_SPI_API_URL);
  if (error) {
    mockedCreateFeed.mockRejectedValue(error);
  } else {
    mockedCreateFeed.mockResolvedValue(MOCK_CREATE_FEED_SUCCESS);
  }
}

function mockFetch(status?: number) {
  if (status) {
    mockRejectFetchResponse(status);
  } else {
    mockResolveFetchResponse(200, {});
  }
}

jest.mock("@/app/sdk/factory/ListingsSPAPIClientsFactory");
const mockedBuildFeedsAPIClient = jest.mocked(buildFeedsAPIClient);
mockedBuildFeedsAPIClient.mockResolvedValue(feedsApi);
const mockedGetSPAPIEndpoint = jest.mocked(getSPAPIEndpoint);

jest.mock("@/app/api/settings/wrapper");
const mockedGetSettings = jest.mocked(getSettings);

function mockSettings(status: number) {
  mockedGetSettings.mockResolvedValue({
    status: status,
    statusText: "Some Status Text",
    settings: MOCK_SETTINGS,
  });
}

function mockGetFeeds(error?: object) {
  mockedGetSPAPIEndpoint.mockReturnValue(MOCK_SPI_API_URL);
  if (error) {
    mockedGetFeeds.mockRejectedValue(error);
  } else {
    mockedGetFeeds.mockResolvedValue(MOCK_GET_FEEDS_RESPONSE);
  }
}

describe("Test for the GET feeds API", () => {
  beforeEach(() => {
    fetch.resetMocks();
    jest.clearAllMocks();
  });

  test("Returns 200 on successful retrieval of feeds", async () => {
    mockSettings(200);
    mockGetFeeds();

    await verifyStatusAndBody(await GET(), 200, true);
  });

  test("returns 500 on settings fetch failure", async () => {
    mockSettings(500);
    await verifyStatusAndBody(await GET(), 500, false);
  });

  test("returns 404 on failure making call to the GetFeeds API", async () => {
    mockSettings(200);
    mockGetFeeds({
      status: 404,
    });
    await verifyStatusAndBody(await GET(), 404, true);
  });
});

describe("Test Create Feed", () => {
  beforeEach(() => {
    fetch.resetMocks();
    jest.clearAllMocks();
  });

  test("Returns 200 on Success full feed creations", async () => {
    mockSettings(200);
    mockCreateFeedDocument();
    mockFetch();
    mockCreateFeed();

    await verifyStatusAndBody(await POST(buildPostNextRequest()), 200, true);
  });

  test("returns 500 on exception from settings response when calling POST Feed", async () => {
    mockSettings(500);

    await verifyStatusAndBody(await POST(buildPostNextRequest()), 500, false);
  });

  test("returns 500 on exception while uploading feed document", async () => {
    mockSettings(200);
    mockCreateFeedDocument();
    mockFetch(500);

    await verifyStatusAndBody(await POST(buildPostNextRequest()), 500, true);
  });

  test("returns 500 on exception while creating feed", async () => {
    mockSettings(200);
    mockCreateFeedDocument();
    mockFetch();
    mockCreateFeed({
      error: 429,
      statusText: "Quota Exceeded",
    });

    await verifyStatusAndBody(await POST(buildPostNextRequest()), 500, true);
  });
});
