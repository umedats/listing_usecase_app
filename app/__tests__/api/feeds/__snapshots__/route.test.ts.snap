// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Create Feed Returns 200 on Success full feed creations 1`] = `
{
  "data": {
    "feedId": "123456789",
  },
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/feeds-api-v2021-06-30-reference#post-feeds2021-06-30documents",
      "apiName": "CreateFeedDocument",
      "request": {
        "contentType": "application/json",
      },
      "response": {
        "feedDocumentId": "123456",
        "url": "https:feed.document/url",
      },
    },
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/feeds-api-v2021-06-30-reference#post-feeds2021-06-30feeds",
      "apiName": "CreateFeed",
      "request": {
        "feedType": "JSON_LISTINGS_FEED",
        "inputFeedDocumentId": "123456",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
      },
      "response": {
        "feedId": "123456789",
      },
    },
  ],
}
`;

exports[`Test Create Feed returns 500 on exception while creating feed 1`] = `
{
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/feeds-api-v2021-06-30-reference#post-feeds2021-06-30documents",
      "apiName": "CreateFeedDocument",
      "request": {
        "contentType": "application/json",
      },
      "response": {
        "feedDocumentId": "123456",
        "url": "https:feed.document/url",
      },
    },
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/feeds-api-v2021-06-30-reference#post-feeds2021-06-30feeds",
      "apiName": "CreateFeed",
      "request": {
        "feedType": "JSON_LISTINGS_FEED",
        "inputFeedDocumentId": "123456",
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
      },
      "response": {
        "error": 429,
        "statusText": "Quota Exceeded",
      },
    },
  ],
}
`;

exports[`Test Create Feed returns 500 on exception while uploading feed document 1`] = `
{
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/feeds-api-v2021-06-30-reference#post-feeds2021-06-30documents",
      "apiName": "CreateFeedDocument",
      "request": {
        "contentType": "application/json",
      },
      "response": {
        "status": 500,
      },
    },
  ],
}
`;

exports[`Test for the GET feeds API Returns 200 on successful retrieval of feeds 1`] = `
{
  "data": [
    {
      "createdTime": "2023-10-05T04:23:58+0000",
      "feedId": "1000",
      "feedType": "JSON_LISTINGS_FEED",
      "processingStatus": "IN_PROGRESS",
      "resultFeedDocumentId": "amzn1.tortuga.3.ed4cd0d8-447b-4c22-96b5-52da8ace1207.T3YUVYPGKE9BMY",
    },
    {
      "createdTime": "2023-10-05T04:23:58+0000",
      "feedId": "2000",
      "feedType": "JSON_LISTINGS_FEED",
      "processingStatus": "IN_PROGRESS",
    },
  ],
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/feeds-api-v2021-06-30-reference#get-feeds2021-06-30feeds",
      "apiName": "GetFeeds",
      "request": {
        "feedTypes": [
          "JSON_LISTINGS_FEED",
        ],
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
        "pageSize": 20,
      },
      "response": {
        "feeds": [
          {
            "createdTime": "2023-10-05T04:23:58+0000",
            "feedId": "1000",
            "feedType": "JSON_LISTINGS_FEED",
            "processingStatus": "IN_PROGRESS",
            "resultFeedDocumentId": "amzn1.tortuga.3.ed4cd0d8-447b-4c22-96b5-52da8ace1207.T3YUVYPGKE9BMY",
          },
          {
            "createdTime": "2023-10-05T04:23:58+0000",
            "feedId": "2000",
            "feedType": "JSON_LISTINGS_FEED",
            "processingStatus": "IN_PROGRESS",
          },
        ],
      },
    },
  ],
}
`;

exports[`Test for the GET feeds API returns 404 on failure making call to the GetFeeds API 1`] = `
{
  "debugContext": [
    {
      "apiDocumentationLink": "https://developer-docs.amazon.com/sp-api/docs/feeds-api-v2021-06-30-reference#get-feeds2021-06-30feeds",
      "apiName": "GetFeeds",
      "request": {
        "feedTypes": [
          "JSON_LISTINGS_FEED",
        ],
        "marketplaceIds": [
          "ATVPDKIKX0DER",
        ],
        "pageSize": 20,
      },
      "response": {
        "status": 404,
      },
    },
  ],
}
`;
